airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_incremental.py

```diff
@@ -170,70 +180,90 @@ async def test_read_sequential_slices(
-            assert state_message.type == Type.STATE
+            # Important!
 
-            # if first state message, skip
-            # this is because we cannot assert if the first state message will result in new records
-            # as in this case it is possible for a connector to return an empty state message when it first starts.
-            # e.g. if the connector decides it wants to let the caller know that it has started with an empty state.
-            if idx == 0:
-                continue
+            # There is only a small subset of assertions we can make
+            # in the absense of enforcing that all connectors return 3 or more state messages
+            # during the first read.
 
-            # if last state message, skip
-            # this is because we cannot assert if the last state message will result in new records
-            # as in this case it is possible for a connector to return a previous state message.
-            # e.g. if the connector is using pagination and the last page is only partially full
-            if idx == len(unique_state_messages) - 1:
+            # To learn more: https://github.com/airbytehq/airbyte/issues/29926
+            if len(unique_state_messages) == 0:
                 continue
 
-            # if batching required, and not a sample, skip
-            if len(unique_state_messages) >= min_batches_to_test and idx % sample_rate != 0:
-                continue
+            if checkpoint_testing_strategy == CheckpointingStrategies.use_state_variation:
+                if len(unique_state_messages) < 3:
+                    logging.warning(f"Skipping stream {stream.stream.name}. Not enough unique state messages to proceed")
+                    continue
+
+            # To avoid spamming APIs we only test a fraction of batches (4 or 5 states by default)
+            min_batches_to_test = 5
+            sample_rate = (len(unique_state_messages) // min_batches_to_test) or 1
+
+            mutating_stream_name_to_per_stream_state = dict()
+            for idx, state_message_data in enumerate(unique_state_messages):
+                state_message, expected_records_count = state_message_data
+                assert state_message.type == Type.STATE
+
+                is_first_state_message = idx == 0
+                is_last_state_message = idx == len(unique_state_messages) - 1
+
+                if checkpoint_testing_strategy == CheckpointingStrategies.use_state_variation:
+                    # if first state message, skip
+                    # this is because we cannot assert if the first state message will result in new records
+                    # as in this case it is possible for a connector to return an empty state message when it first starts.
+                    # e.g. if the connector decides it wants to let the caller know that it has started with an empty state.
+                    if is_first_state_message:
```

Comment:
doesn't this edge case also apply to the case where we only rely on the last state message? why do we only skip the state message with the `use_state_variation` but know to expect records otherwise?

Reply:
When we rely on the last state message, we don't need to calculate first, last, or rate to decide whether to use the state or not. We have only one state message for testing, which simultaneously satisfies the first, last, and rate conditions